File: wireless_watcher/wireless_watcher/wireless_watcher.py
#!/usr/bin/env python3
# Software License Agreement (BSD)
#
# @author    Mike Purvis <mpurvis@clearpath.ai>
# @author    Roni Kreinin <rkreinin@clearpathrobotics.com>
# @copyright (c) 2022, Clearpath Robotics, Inc., All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification, are permitted provided that
# the following conditions are met:
# * Redistributions of source code must retain the above copyright notice, this list of conditions and the
#   following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
#   following disclaimer in the documentation and/or other materials provided with the distribution.
# * Neither the name of Clearpath Robotics nor the names of its contributors may be used to endorse or
#   promote products derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
# PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
# TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import os
import re
import rclpy
import time
import threading
import std_msgs.msg
import subprocess

from wireless_msgs.msg import Connection, Network

from rclpy.qos import qos_profile_sensor_data

SYS_NET_PATH = '/sys/class/net'


def trim(field, type=int):
    m = re.match('[0-9-.]+', field)
    if m:
        return type(m.group(0))


class WirelessConnection(Connection):
    def __init__(self, fields):
        args = {}

        args['bitrate'] = trim(fields.get('Bit Rate', "0.0"), float)
        args['txpower'] = trim(fields.get('Tx-Power', "0"))
        args['signal_level'] = trim(fields.get('Signal level', "0"))
        args['noise_level'] = trim(fields.get('Noise level', "0"))
        args['essid'] = fields.get('ESSID', "").strip('"')
        args['bssid'] = fields.get('Access Point', "").strip()
        args['frequency'] = trim(fields.get('Frequency', "0.0"), float)

        try:
            args['link_quality_raw'] = fields['Link Quality']
            num, den = re.split('/', args['link_quality_raw'])
            args['link_quality'] = float(num) / float(den)
        except:
            pass

        self.msg = Connection(**args)


class WirelessNetwork(Network):
    def __init__(self, fields):
        args = {}
        self.msg = Network(**args)


def main():
    rclpy.init()

    node = rclpy.create_node('wireless_watcher')

    node.declare_parameter('hz', 1)
    node.declare_parameter('dev', '')
    node.declare_parameter('connected_topic', 'connected')
    node.declare_parameter('connection_topic', 'connection')

    hz = node.get_parameter('hz').value
    dev = node.get_parameter('dev').value
    connected_topic = node.get_parameter('connected_topic').value
    connection_topic = node.get_parameter('connection_topic').value

    if not dev:
        wldevs = [d for d in os.listdir(SYS_NET_PATH) if d.startswith('wl') or d.startswith('wifi')]
        if wldevs:
            dev = wldevs[0]
        else:
            node.get_logger().fatal('No wireless device found to monitor.')
            return 1

    node.get_logger().info('Monitoring {0}'.format(dev))

    previous_error = False
    previous_success = False

    thread = threading.Thread(target=rclpy.spin, args=(node, ), daemon=True)
    thread.start()

    connected_pub = node.create_publisher(std_msgs.msg.Bool, connected_topic, qos_profile_sensor_data)
    connection_pub = node.create_publisher(Connection, connection_topic, qos_profile_sensor_data)

    # Disable this until we actually collect and publish the data.
    # network_pub = node.create_publisher(Network, 'network', qos_profile_sensor_data)

    connected_msg = std_msgs.msg.Bool()

    while rclpy.ok():
        try:
            with open(os.path.join(SYS_NET_PATH, dev, 'operstate'), 'rb') as f:
                connected_msg.data = f.read().strip() == b'up'
        except FileNotFoundError:
            connected_msg.data = False
        finally:
            connected_pub.publish(connected_msg)

        try:
            wifi_str = subprocess.check_output(['iwconfig', dev], stderr=subprocess.STDOUT).decode()
            fields_str = re.split('\s\s+', wifi_str)
            fields_list = [re.split('[:=]', field_str, maxsplit=1) for field_str in fields_str]
            fields_dict = {'dev': fields_str[0], 'type': fields_str[1]}
            fields_dict.update(dict([field for field in fields_list if len(field) == 2]))

            # Check if WiFi is connected to a network
            if "Not-Associated" not in fields_dict['Access Point']:
                connection = WirelessConnection(fields_dict)
                connection_pub.publish(connection.msg)

            if not previous_success:
                previous_success = True
                previous_error = False
                node.get_logger().info('Retrieved status of interface {0}. Now updating at {1} Hz.'.format(dev, hz))

        except subprocess.CalledProcessError:
            if not previous_error:
                previous_error = True
                previous_success = False
                node.get_logger().error('Error checking status of interface {0}. Will try again at {1} Hz.'.format(dev, hz))

        time.sleep(1.0 / hz)

    thread.join()
    return 0


if __name__ == '__main__':
    import sys
    sys.exit(main())

File: wireless_watcher/src/wireless_watcher.cpp
/**
 *
 *  \file
 *  \brief      Wireless Watcher Node
 *  \author     Mike Purvis <mpurvis@clearpathrobotics.com>
 *  \author     Roni Kreinin <rkreinin@clearpathrobotics.com>
 *  \author     Tony Baltovski <tbaltovski@clearpathrobotics.com>
 *  \copyright  Copyright (c) 2023, Clearpath Robotics, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of Clearpath Robotics, Inc. nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL CLEARPATH ROBOTICS, INC. BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * Please send comments, questions, or patches to code@clearpathrobotics.com
 *
 */

#include "wireless_watcher/wireless_watcher.hpp"

#include <arpa/inet.h>
#include <chrono>
#include <cstdio>
#include <dirent.h>
#include <fstream>
#include <ifaddrs.h>
#include <iostream>
#include <limits>
#include <memory>
#include <regex>
#include <string>
#include <thread>
#include <unordered_map>
#include <vector>

#include "diagnostic_updater/diagnostic_updater.hpp"

using namespace std::chrono_literals;

WirelessWatcher::WirelessWatcher()
    : rclcpp::Node("wireless_watcher"), updater_(this) {
  this->declare_parameter("hz", 1.0);
  this->declare_parameter("dev", "");
  this->declare_parameter("connected_topic", "connected");
  this->declare_parameter("connection_topic", "connection");

  hz = this->get_parameter("hz").as_double();
  dev = this->get_parameter("dev").as_string();
  connected_topic = this->get_parameter("connected_topic").as_string();
  connection_topic = this->get_parameter("connection_topic").as_string();

  if (dev.empty()) {
    std::vector<std::string> wldevs;
    DIR *dir;
    struct dirent *ent;
    if ((dir = opendir(SYS_NET_PATH)) != NULL) {
      while ((ent = readdir(dir)) != NULL) {
        std::string dev_name = ent->d_name;
        if (dev_name.compare(0, 2, "wl") == 0 ||
            dev_name.compare(0, 4, "wifi") == 0) {
          wldevs.push_back(dev_name);
        }
      }
      closedir(dir);
    } else {
      RCLCPP_FATAL(this->get_logger(), "Failed to open directory: %s",
                   SYS_NET_PATH);
      return;
    }
    if (!wldevs.empty()) {
      dev = wldevs[0];
    } else {
      RCLCPP_FATAL(this->get_logger(), "No wireless device found to monitor.");
      return;
    }
  }

  RCLCPP_INFO(this->get_logger(), "Monitoring %s", dev.c_str());

  connected_pub_ = this->create_publisher<std_msgs::msg::Bool>(
      connected_topic, rclcpp::SensorDataQoS());
  connection_pub_ = this->create_publisher<wireless_msgs::msg::Connection>(
      connection_topic, rclcpp::SensorDataQoS());

  timer_ = this->create_wall_timer(
      std::chrono::milliseconds(static_cast<int>(1000.0 / hz)),
      std::bind(&WirelessWatcher::timer_callback, this));

  updater_.setHardwareID(dev);
  updater_.add("Wi-Fi Monitor", this, &WirelessWatcher::diagnostic);
}

void WirelessWatcher::timer_callback() {
  try {
    std::string operstate_filepath = std::string(SYS_NET_PATH);
    operstate_filepath += "/";
    operstate_filepath += dev;
    operstate_filepath += "/operstate";
    std::ifstream operstate_file(operstate_filepath.c_str());
    std::string operstate;
    operstate_file >> operstate;
    connected_msg_.data = operstate == "up";
  } catch (const std::exception &e) {
    connected_msg_.data = false;
  }
  connected_pub_->publish(connected_msg_);

  if (!connected_msg_.data) {
    return;
  }

  std::string iw_link = exec_cmd("iw dev " + dev + " link 2>/dev/null");

  if (iw_link.find("Not connected") != std::string::npos || iw_link.empty()) {
    return;
  }

  std::smatch m;

  // BSSID: "Connected to xx:xx:xx:xx:xx:xx (on wlan0)"
  if (std::regex_search(iw_link, m,
                        std::regex(R"(Connected to ([\da-f:]+))"))) {
    connection_msg_.bssid = m[1].str();
  }

  // SSID
  if (std::regex_search(iw_link, m, std::regex(R"(\bSSID: (.+))"))) {
    std::string s = m[1].str();
    s.erase(s.find_last_not_of(" \t\r\n") + 1);
    connection_msg_.essid = s;
  }

  // freq in MHz, convert to GHz
  if (std::regex_search(iw_link, m, std::regex(R"(\bfreq: (\d+))"))) {
    connection_msg_.frequency = std::stod(m[1].str()) / 1000.0;
  }

  // signal in dBm
  if (std::regex_search(iw_link, m, std::regex(R"(\bsignal: (-?\d+) dBm)"))) {
    connection_msg_.signal_level = static_cast<int16_t>(std::stoi(m[1].str()));
  }

  // tx bitrate in MBit/s
  if (std::regex_search(iw_link, m,
                        std::regex(R"(\btx bitrate: ([\d.]+) MBit/s)"))) {
    try {
      connection_msg_.bitrate = std::stof(m[1].str());
    } catch (const std::invalid_argument &) {
      connection_msg_.bitrate = std::numeric_limits<float>::quiet_NaN();
    }
  } else {
    connection_msg_.bitrate = std::numeric_limits<float>::quiet_NaN();
  }

  // txpower from iw dev info
  std::string iw_info = exec_cmd("iw dev " + dev + " info 2>/dev/null");
  if (std::regex_search(iw_info, m, std::regex(R"(\btxpower ([\d.]+) dBm)"))) {
    connection_msg_.txpower = static_cast<int16_t>(std::stof(m[1].str()));
  }

  // Derive link quality from signal level (-100 dBm = 0%, -50 dBm = 100%)
  int quality_pct;
  if (connection_msg_.signal_level >= -50) {
    quality_pct = 100;
  } else if (connection_msg_.signal_level <= -100) {
    quality_pct = 0;
  } else {
    quality_pct = 2 * (connection_msg_.signal_level + 100);
  }
  connection_msg_.link_quality_raw = std::to_string(quality_pct) + "/100";
  connection_msg_.link_quality = static_cast<float>(quality_pct) / 100.0f;

  connection_pub_->publish(connection_msg_);
}

std::string WirelessWatcher::exec_cmd(const std::string &cmd) {
  std::array<char, 128> buffer;
  std::string result;
  std::unique_ptr<FILE, int (*)(FILE *)> pipe(popen(cmd.c_str(), "r"), pclose);
  if (!pipe) {
    throw std::runtime_error("popen() failed!");
  }
  while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
    result += buffer.data();
  }
  return result;
}

std::vector<std::string> WirelessWatcher::split(const std::string &s,
                                                const std::string &delimiter) {
  std::regex regex(delimiter);
  std::sregex_token_iterator it(s.begin(), s.end(), regex, -1);
  std::vector<std::string> tokens{it, {}};
  return tokens;
}

void WirelessWatcher::diagnostic(
    diagnostic_updater::DiagnosticStatusWrapper &stat) {
  stat.add("Wireless Network Interface", dev);
  stat.add("Wi-Fi Connected", connected_msg_.data ? "True" : "False");

  if (!connected_msg_.data) {
    stat.summaryf(diagnostic_updater::DiagnosticStatusWrapper::WARN,
                  "%s Disconnected", dev.c_str());
    return;
  }
  stat.summary(diagnostic_updater::DiagnosticStatusWrapper::OK, "OK");

  ip_address_diag(dev, stat);
  stat.add("Frequency (GHz)", connection_msg_.frequency);
  stat.add("ESSID", connection_msg_.essid);
  stat.add("BSSID", connection_msg_.bssid);
  stat.add("Transmit Power (dBm)", connection_msg_.txpower);
  stat.add("Theoretical Max Bitrate (Mbps)", connection_msg_.bitrate);
  stat.add("Link Quality Raw", connection_msg_.link_quality_raw);
  stat.addf("Link Quality (%)", "%.1f", connection_msg_.link_quality * 100);
  stat.add("Signal Strength (dBm)", connection_msg_.signal_level);

  if (connection_msg_.signal_level < SIGNAL_STRENGTH_VERY_WEAK) {
    stat.mergeSummaryf(diagnostic_updater::DiagnosticStatusWrapper::WARN,
                       "Very Poor Signal Strength (%d dBm)",
                       connection_msg_.signal_level);
  } else if (connection_msg_.signal_level < SIGNAL_STRENGTH_WEAK) {
    stat.mergeSummaryf(diagnostic_updater::DiagnosticStatusWrapper::WARN,
                       "Poor Signal Strength (%d dBm)",
                       connection_msg_.signal_level);
  }
}

void WirelessWatcher::ip_address_diag(
    std::string dev, diagnostic_updater::DiagnosticStatusWrapper &stat) {

  struct ifaddrs *ptr_ifaddrs = nullptr, *entry;

  if (getifaddrs(&ptr_ifaddrs) == 0) {
    for (entry = ptr_ifaddrs; entry != nullptr; entry = entry->ifa_next) {
      // Find the requested interface and ensure it has an address
      if (std::string(entry->ifa_name) != dev || entry->ifa_addr == nullptr) {
        continue;
      }

      sa_family_t address_family = entry->ifa_addr->sa_family;
      // Skip if the address is not IPv4
      if (address_family != AF_INET) {
        continue;
      }
      char buffer[INET_ADDRSTRLEN] = {};
      inet_ntop(address_family,
                &((struct sockaddr_in *)(entry->ifa_addr))->sin_addr, buffer,
                INET_ADDRSTRLEN);

      stat.add("IP Address", std::string(buffer));

      if (entry->ifa_netmask != nullptr) {
        char buffer[INET_ADDRSTRLEN] = {
            0,
        };
        inet_ntop(address_family,
                  &((struct sockaddr_in *)(entry->ifa_netmask))->sin_addr,
                  buffer, INET_ADDRSTRLEN);

        stat.add("Netmask", std::string(buffer));
      } else {
        stat.add("Netmask", "Not found");
      }

      freeifaddrs(ptr_ifaddrs);
      return;
    }
  }
  stat.mergeSummaryf(diagnostic_updater::DiagnosticStatusWrapper::WARN,
                     "Failed to get IP addresses for %s", dev.c_str());
  stat.add("IP Address", "Not found");
  stat.add("Netmask", "Not found");

  if (ptr_ifaddrs != nullptr) {
    freeifaddrs(ptr_ifaddrs);
  }
  return;
}

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<WirelessWatcher>());
  rclcpp::shutdown();
  return 0;
}
